Adding Terraform Enterprise security model (#1763)
hashicorp/terraform-website: content/source/docs/enterprise/system-overview/security-model.html.md

### Site Admin

[Site admins](../admin/admin-access.html) are responsible for application-level configuration of Terraform Enterprise. Via the admin interface they will be able to manage all users, workspaces and organizations and will have access to all data stored within Terraform Enterprise. Site admins are also responsible for configuring SAML and will be the only users still able to access Terraform Enterprise via username/password once SAML is configured. 

Due to the extensive capabilities granted to this role, the site admin role should be restricted to a small number of users within your organization.

## Differences Between Terraform Enterprise and Terraform Cloud Security Models

For the most part, Terraform Enterprise’s threat model can be considered a superset of the [Terraform Cloud security model](../../cloud/architectural-details/security-model.html) and (with the exception of the points listed below) all of the content on that page applies to Terraform Enterprise. Terraform Enterprise’s threat model differs from that of Terraform Clouds in these respects:

### TFE Requires You to Manage and Secure the Underlying Network and Infrastructure

> TFE infrastructure admins are required to manage all aspects of the underlying infrastructure including initial provisioning, secure configuration, access control, configuring network ACLs, and OS-level software updates. TFE can not ensure the security of your data if the underlying infrastructure is compromised.

Reviewer: ```suggestion
Infrastructure admins are required to manage all aspects of the underlying infrastructure. This includes initial provisioning, secure configuration, access control, network ACL configuration, and OS-level software updates. Terraform Enterprise cannot ensure the security of your data if the underlying infrastructure is compromised.
```

> ### You are Responsible for Updating Your TFE Deployment

Reviewer: ```suggestion
### You are Responsible for Updating Your Terraform Enterprise Deployment
```

> Security fixes are released along with application features and bug fixes via TFE’s standard monthly release cycle. TFE infrastructure admins are responsible for applying updates.

Reviewer: I'm also wondering - is there a release notes page/section that we could link to here?

```suggestion
We release security fixes, application features, and bug fixes for Terraform Enterprise each month. Infrastructure admins are responsible for applying updates.
```

> ### You are Responsible for TFE’s Availability, Backups, and Disaster Recovery

Reviewer: ```suggestion
### You are Responsible for Availability, Backups, and Disaster Recovery
```

> TFE infrastructure admins are responsible for all aspects of TFE’s reliability and availability. Refer to TFE documentation on [monitoring](../admin/monitoring.html), [backups and restores](../admin/backup-restore.html), and [high availability mode (active/active)](../admin/active-active.html) for more guidance on this topic.

Reviewer: ```suggestion
Infrastructure admins are responsible for all aspects of reliability and availability. Refer to Terraform Enterprise documentation on [monitoring](../admin/monitoring.html), [backups and restores](../admin/backup-restore.html), and [high availability mode (active/active)](../admin/active-active.html) for more guidance on this topic.
```

> ### TFE Isolates Terraform Operations via Docker Containers, not a separate environment

Reviewer: ```suggestion
### Terraform Enterprise Isolates Terraform Operations via Docker Containers
```

> Unlike Terraform Cloud, TFE executes all Terraform operations in Docker containers on the TFE host. The containers are assigned to an isolated Docker network to prevent them from communicating with TFE’s backend services, but may have access to resources available on hosts accessible from the your TFE instance. 

Reviewer: ```suggestion
Unlike Terraform Cloud, Terraform Enterprise executes all Terraform operations in Docker containers on the Terraform Enterprise host. The containers are assigned to an isolated Docker network to prevent them from communicating with Terraform Enterprise backend services, but may have access to resources available on hosts accessible from the your Terraform Enterprise instance. 
```

## Recommendations for Securely Operating Terraform Enterprise

> In addition the the recommendations provided in the [Terraform Cloud security model](../../cloud/architectural-overview/security-model.html), we offer these recommendations to users of Terraform Enterprise 

Reviewer: ```suggestion
In addition those provided in the [Terraform Cloud security model](../../cloud/architectural-overview/security-model.html), we recommend the following for Terraform Enterprise users. 
```

### Run Terraform Enterprise in an Isolated Network, Limit Ingress Ports, and Restrict Access to Underlying Infrastructure

> To minimize attack surface, we recommend running TFE in an isolated network and limiting ingress ports to only 80 and 443. 

Reviewer: ```suggestion
To minimize attack surface, we recommend running Terraform Enterprise in an isolated network and limiting ingress ports to only 80 and 443. 
```

> For standalone deployments, port 8800 is reserved for the [Replicated admin console](../admin/admin-access.html), which is used for configuring TFE. This port should only be exposed to TFE infrastructure admins. Alternatively, if you choose to configure TFE via the [automated process](../install/automating-the-installer.html), you can disable the Replicated admin console by passing the `disable-replicated-ui` argument to the installation script:

Reviewer: ```suggestion
For standalone deployments, port 8800 is reserved for the [Replicated admin console](../admin/admin-access.html), which is used for configuring Terraform Enterprise. This port should only be exposed to infrastructure admins. If you choose to configure Terraform Enterprise with the [automated process](../install/automating-the-installer.html), you can disable the Replicated admin console by passing the `disable-replicated-ui` argument to the installation script:
```

```sudo bash ./install.sh disable-replicated-ui```

> Additionally, we recommend restricting access to the nodes that are running TFE. TFE can not ensure the security or integrity of its data if the underlying infrastructure is compromised.

Reviewer: ```suggestion
Additionally, we recommend restricting access to the nodes that are running Terraform Enterprise. Terraform Enterprise can not ensure the security or integrity of its data if the underlying infrastructure is compromised.
```

### Enable Optional Security Features

Once you are ready to use Terraform Enterprise for production workloads, we recommend enabling these optional security features.

> #### Strict Transport Security Header

Reviewer: Starting with a verb to match others in this section.
```suggestion
#### Set Strict Transport Security Header
```

> You can configure TFE to set the [Strict Transport Security (HSTS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security) header via the Settings page of the installer dashboard by enabling “Force TLS” radio button under “SSL/TLS Configuration”

Reviewer: ```suggestion
You can configure Terraform Enterprise to set the [Strict Transport Security (HSTS)](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security) header by:
* Visiting the installer dashboard "Settings" page and enabling “Force TLS” under the “SSL/TLS Configuration” section.
* Setting [force_tls](../install/automating-the-installer.html#force_tls) in the application settings file.
```

> You can also enable this setting via the application settings file file with the [force_tls](../install/automating-the-installer.html#force_tls) setting.

Reviewer: Remove as I added it back in above as part of the list.
```suggestion
```

> ~> **Note:** Once properly configured, the HSTS header cannot be disabled and will prevent clients from accessing your TFE domain via HTTP or HTTPS using a self-signed cert. We recommend only enabling this setting for production TFE deployments.

Reviewer: ```suggestion
~> **Note:** Once properly configured, the HSTS header cannot be disabled and will prevent clients from accessing your Terraform Enterprise domain via HTTP or HTTPS using a self-signed cert. We recommend only enabling this setting for production Terraform Enterprise deployments.
```

#### Restrict Terraform Build Worker Metadata Access

> By default Terraform Enterprise does not prevent Terraform operations from accessing the instance metadata service, which may contain IAM credentials or other sensitive data. Refer to [AWS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html), [Azure](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows), or [Google Cloud's](https://cloud.google.com/compute/docs/storing-retrieving-metadata) documentation for more information on this service.

Reviewer: ```suggestion
By default, Terraform Enterprise does not prevent Terraform operations from accessing the instance metadata service, which may contain IAM credentials or other sensitive data. Refer to [AWS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html), [Azure](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service?tabs=windows), or [Google Cloud](https://cloud.google.com/compute/docs/storing-retrieving-metadata) documentation for more information on this service.
```

> TFE allows you to restrict access to the metadata endpoint from Terraform operations, preventing workspaces from reading any data from the metadata service. This feature can be enabled via the “Restrict Terraform Build Worker Instance Metadata Access” setting under the “Advanced Configuration” section in the installer dashboard’s settings page, or via the [restrict_worker_metadata_access](../install/automating-the-installer.html#restrict_worker_metadata_access) setting in the application settings file.

Reviewer: ```suggestion
Terraform Enterprise allows you to restrict access to the metadata endpoint from Terraform operations, preventing workspaces from reading any data from the metadata service. You can do this by:
* Visiting the installer dashboard "Settings" page and enabling “Restrict Terraform Build Worker Instance Metadata Access” under the “Advanced Configuration” section. 
* Setting [restrict_worker_metadata_access](../install/automating-the-installer.html#restrict_worker_metadata_access) in the application settings file.
```

> Unless you are relying on the [instance profile as a means of providing default credentials to TFE workspaces](../before-installing/index.html#aws-specific-configuration), we recommend enabling this setting to prevent Terraform operations from accessing the instance metadata endpoint. 

Reviewer: ```suggestion
We recommend enabling this setting to prevent Terraform operations from accessing the instance metadata endpoint, unless you are relying on the [instance profile to provide default credentials to workspaces](../before-installing/index.html#aws-specific-configuration). 
```

> #### Ensure Global Remote State Sharing is Disabled

Reviewer: ```suggestion
#### Disable Global Remote State Sharing
```

> TFE allows administrators to enable [global remote state sharing](../admin/general.html#remote-state-sharing), a feature that allows any workspace to access the state versions of any other workspace within the same organization. We recommend disabling the feature and relying on [controlled remote state access](https://www.hashicorp.com/blog/announcing-controlled-remote-state-access-for-terraform-cloud-and-enterprise) if you need to share state between workspaces.

Reviewer: Also - wondering which type of admin - there are two listed above. Maybe we should specify if it's both or just one. 
```suggestion
Terraform Enterprise allows administrators to enable [global remote state sharing](../admin/general.html#remote-state-sharing), which allows any workspace to access the state versions of any other workspace within the same organization. We recommend disabling this feature and relying on [controlled remote state access](https://www.hashicorp.com/blog/announcing-controlled-remote-state-access-for-terraform-cloud-and-enterprise) if you need to share state between workspaces.
```

Author: I'm referring to site admins here. I will update that to be more clear

#### Treat Support Bundles with Care

> Support bundles are used to share diagnostic information with HashiCorp support. Please note that support bundles may contain sensitive information from your Terraform Enterprise installation. You should consider these bundles sensitive; they should not be shared with untrusted parties and should be deleted as soon as possible.

Reviewer: ```suggestion
Terraform Enterprise uses support bundles to share diagnostic information with HashiCorp support. Please note that support bundles may contain sensitive information from your Terraform Enterprise installation. You should not share them with untrusted parties and should delete them as soon as possible.
```

#### Regularly Update Terraform Enterprise

> Terraform Enterprise updates are released on a monthly cadence. Updates may contain additional security features or fixes for existing security vulnerabilities, so we recommend establishing a process for periodically updating your Terraform Enterprise installation.

Reviewer: ```suggestion
We release Terraform Enterprise updates each month. Updates may contain additional security features or fixes for existing security vulnerabilities, so we recommend establishing a process for periodically updating your Terraform Enterprise installation.
```

#### Subscribe to Terraform Enterprise Security Bulletins

> HashiCorp publishes security updates, which address security vulnerabilities in HashiCorp products, in the Security category of [HashiCorp Discuss](https://discuss.hashicorp.com/c/security/).

Reviewer: ```suggestion
We publish security updates that address security vulnerabilities in HashiCorp products. You can find them in the Security category of [HashiCorp Discuss](https://discuss.hashicorp.com/c/security/).
```

> We recommend that Terraform Enterprise administrators follow the [documented steps](https://discuss.hashicorp.com/t/about-hashicorp-security-updates/15330) to subscribe to email notifications or the RSS feed for Terraform Enterprise security updates.

Reviewer: both infrastructure and site admins?

Author: Infrastructure admins - they are the ones responsible for updating Terraform Enterprise.

---
layout: "enterprise"
page_title: "Security Model- System Overview - Terraform Enterprise"
---

# Terraform Enterprise Security Model

> ## Purpose of this document

Reviewer: I think we can omit this and just jump right into the explanation :) 
```suggestion
```

### Terraform Enterprise Requires You to Manage and Secure the Underlying Network and Infrastructure

Infrastructure admins are required to manage all aspects of the underlying infrastructure. This includes initial provisioning, secure configuration, access control, network ACL configuration, and OS-level software updates. Terraform Enterprise cannot ensure the security of your data if the underlying infrastructure is compromised.

### You are Responsible for Updating Your Terraform Enterprise Deployment

We release security fixes, application features, and bug fixes for Terraform Enterprise each month. Infrastructure admins are responsible for applying updates.

### You are Responsible for Availability, Backups, and Disaster Recovery

Infrastructure admins are responsible for all aspects of reliability and availability. Refer to Terraform Enterprise documentation on [monitoring](../admin/monitoring.html), [backups and restores](../admin/backup-restore.html), and [high availability mode (active/active)](../admin/active-active.html) for more guidance on this topic.

### Terraform Enterprise Isolates Terraform Operations via Docker Containers

> Unlike Terraform Cloud, Terraform Enterprise executes all Terraform operations in Docker containers on the Terraform Enterprise host. The containers are assigned to an isolated Docker network to prevent them from communicating with Terraform Enterprise backend services, but may have access to resources available on hosts accessible from the your Terraform Enterprise instance. 

Reviewer: ```suggestion
Unlike Terraform Cloud, Terraform Enterprise executes all Terraform operations in Docker containers on the Terraform Enterprise host. The containers are assigned to an isolated Docker network to prevent them from communicating with Terraform Enterprise backend services, but may have access to host resources accessible from your Terraform Enterprise instance. 
```

Author: Hmm... this one is a little tricky. I'm not exactly sure how to phrase this. The point I'm trying to make here is that Terraform runs within TFE will have access to anything else accessible on the network - any other web services, databases, the instance metadata URL, etc. TFE doesn't do any sort of egress filtering on Terraform runs.

In retrospect, maybe this point is kind of obvious? One of the value propositions of TFE is that you can use it manage private internal resources - internal databases, services, etc.

Maybe we should just remove everything after the comma:
```
Unlike Terraform Cloud, Terraform Enterprise executes all Terraform operations in Docker containers on the Terraform Enterprise host. The containers are assigned to an isolated Docker network to prevent them from communicating with Terraform Enterprise backend services.
```

Reviewer: got it! I trust your judgment about whether we need to say that again - I think it's much cleaner looking now, but if you think we should doubly emphasize, I wonder if you could say 

_Unlike Terraform Cloud, Terraform Enterprise performs all Terraform operations in Docker containers on the Terraform Enterprise host. The containers are assigned to an isolated Docker network to prevent them from communicating with Terraform Enterprise backend services. However, Terraform Enterprise does not perform any egress filtering, so Terraform runs can still access available network resources like web services and databases._

## Recommendations for Securely Operating Terraform Enterprise

In addition those provided in the [Terraform Cloud security model](../../cloud/architectural-overview/security-model.html), we recommend the following for Terraform Enterprise users. 

### Run Terraform Enterprise in an Isolated Network, Limit Ingress Ports, and Restrict Access to Underlying Infrastructure

To minimize attack surface, we recommend running Terraform Enterprise in an isolated network and limiting ingress ports to only 80 and 443. 

For standalone deployments, port 8800 is reserved for the [Replicated admin console](../admin/admin-access.html), which is used for configuring Terraform Enterprise. This port should only be exposed to infrastructure admins. If you choose to configure Terraform Enterprise with the [automated process](../install/automating-the-installer.html), you can disable the Replicated admin console by passing the `disable-replicated-ui` argument to the installation script:

> ```sudo bash ./install.sh disable-replicated-ui```

Reviewer: I worry because it shows the rest of this page like it's going to be formatted as code (or at least differently than text). Please just test this locally before pushing to verify that it will render properly :)

Author: Tested, it looks good!

> Additionally, we recommend restricting access to the nodes that are running Terraform Enterprise. Terraform Enterprise can not ensure the security or integrity of its data if the underlying infrastructure is compromised.

Reviewer: We've said this above. Keep this if you think it's worth saying again (it could be valuable to emphasize), but then please just change "its data" to "your data".

```suggestion
Additionally, we recommend restricting access to the nodes that are running Terraform Enterprise.
```

Author: I think its valuable to emphasize this. I'll update it to say "your data"

We recommend enabling this setting to prevent Terraform operations from accessing the instance metadata endpoint, unless you are relying on the [instance profile to provide default credentials to workspaces](../before-installing/index.html#aws-specific-configuration). 

#### Disable Global Remote State Sharing

Terraform Enterprise allows site admins to enable [global remote state sharing](../admin/general.html#remote-state-sharing), which allows any workspace to access the state versions of any other workspace within the same organization. We recommend disabling this feature and relying on [controlled remote state access](https://www.hashicorp.com/blog/announcing-controlled-remote-state-access-for-terraform-cloud-and-enterprise) if you need to share state between workspaces.

#### Treat Support Bundles with Care

Terraform Enterprise uses support bundles to share diagnostic information with HashiCorp support. Please note that support bundles may contain sensitive information from your Terraform Enterprise installation. You should not share them with untrusted parties and should delete them as soon as possible.

> #### Regularly Update Terraform Enterprise

Reviewer: a nit to make sure this one starts with a verb too :-) 

```suggestion
#### Update Terraform Enterprise Often
```

We release Terraform Enterprise updates each month. Updates may contain additional security features or fixes for existing security vulnerabilities, so we recommend establishing a process for periodically updating your Terraform Enterprise installation.

#### Subscribe to Terraform Enterprise Security Bulletins

> We publish security updates that address security vulnerabilities in HashiCorp products. You can find them in the Security category of [HashiCorp Discuss](https://discuss.hashicorp.com/c/security/).

Reviewer: Just removing the first security to help it flow better. This is a nit :) 

```suggestion
We publish updates that address security vulnerabilities in HashiCorp products. You can find them in the Security category of [HashiCorp Discuss](https://discuss.hashicorp.com/c/security/).
```
